### ISSUE TITLE
Add the ARM review labels only after NewAPIVersionRequired is addressed

### ISSUE THREAD

**Issue by bdefoy** [Azure/azure-sdk-tools#6076]
As a part of the automation, it should be made clear that the author needs to follow up with the `NewAPIVersionRequired` action label to create a new API version before moving on to the ARM API review. Only after the `NewAPIVersionRequired` label is addressed should the `WaitForARMfeedback` label be added, making the PR actionable for the ARM reviewers.

Additionally, when the `NewApiVersionRequired` label is added by automation, it should remove the `WaitForARMFeedback` label, add the `ARMChangesRequested` label, and add a comment for the PR Author to remove the `ARMChangesRequested` label once they have created a new API version or obtained an exception from the Breaking Changes Review Board.

**Comment by konrad-jamrozik** [comment]
@bdefoy I have some clarifying questions.

Re:

> As a part of the automation, it should be made clear that the author needs to follow up with the NewAPIVersionRequired action label to create a new API version before moving on to the ARM API review.  Only after the NewAPIVersionRequired label is addressed should the WaitForARMfeedback label be added, making the PR actionable for the ARM reviewers. 

If I understand correctly, you are asking for it to be made clearer to the PR submitter what is the meaning of the `NewAPIVersionRequired` label? I understand that label should only ever by added by our automation, not by hand. By the breaking change detectors we have, to be exact (I grokked the code to confirm this). I guess the ask here is to add some GitHub comment to the PR saying what are the steps to address this label, correct? Like, how to create the new API version and then to request the ARM feedback by adding `WaitForARMfeedback` label.

> Additionally, when the NewApiVersionRequired label is added by automation, it should remove the WaitForARMFeedback label, add the ARMChangesRequested label, 

So I understand correctly the idea here is that as long as the `NewApiVersionRequired` label is in effect, there also has to be `ARMChangesRequested` label in effect. Basically the `NewApiVersionRequired` is a sub-type of all possible changes requested by ARM. Correct?

And you also don't want to see the `WaitForARMFeedback` label because it tells your folks to do manual review, but you don't want to waste reviewing the code until a new API version has been properly defined (or the PR submitter got an exemption, so that your reviewers know it is OK to review the existing API version).

> and add a comment for the PR Author to remove the ARMChangesRequested label once they have created a new API version or obtained an exception from the Breaking Changes Review Board.

OK, so this is basically part of the instructions the PR submitter should follow when addressing the `NewAPIVersionRequired`. The instructions that should be made clear, as described above, in the first quote?

**Comment by bdefoy** [comment]
> > As a part of the automation, it should be made clear that the author needs to follow up with the NewAPIVersionRequired action label to create a new API version before moving on to the ARM API review.  Only after the NewAPIVersionRequired label is addressed should the WaitForARMfeedback label be added, making the PR actionable for the ARM reviewers.
> 
> If I understand correctly, you are asking for it to be made clearer to the PR submitter what is the meaning of the `NewAPIVersionRequired` label, and relevant action items? I understand that label should only ever by added by our automation, not by hand. By the breaking change detectors we have, to be exact (I grokked the code to confirm this). I guess the ask here is to add some GitHub comment to the PR saying what are the steps to address this label, correct? Like, how to create the new API version and then to request the ARM feedback by adding `WaitForARMfeedback` label?

It looks like we already have a bot comment for this label. E.g., https://github.com/Azure/azure-rest-api-specs/pull/24010#issuecomment-1549486347. The comment mentions that a new API version should be added and why, but it should also tell the PR author that the ARM review will not continue until that is addressed. Also, yes, they should need to add the `WaitForARMFeedback` label to continue the review. Adding this label will remove the `ARMChangesRequested` label. This requirement is already mentioned in a comment when the `ARMChangesRequested` label is added though. E.g., https://github.com/Azure/azure-rest-api-specs-pr/pull/12565#issuecomment-1550381726.

> > Additionally, when the NewApiVersionRequired label is added by automation, it should remove the WaitForARMFeedback label, add the ARMChangesRequested label,
> 
> If I understand correctly, the idea here is that as long as the `NewApiVersionRequired` label is in effect, there also has to be `ARMChangesRequested` label in effect. Basically the `NewApiVersionRequired` is a sub-type of all possible changes requested by ARM. Correct?

Yes, `NewApiVersionRequired` is a change requested by ARM, and when it is active on a PR we want the `ARMChangesRequested` label to be active as well.

> And you also don't want to see the `WaitForARMFeedback` label because it tells your folks to do manual review, but you don't want to waste reviewing the code until a new API version has been properly defined (or the PR submitter got an exemption, so that your reviewers know it is OK to review the existing API version).

Exactly.

> > and add a comment for the PR Author to remove the ARMChangesRequested label once they have created a new API version or obtained an exception from the Breaking Changes Review Board.
> 
> OK, so this is basically part of the instructions the PR submitter should follow when addressing the `NewAPIVersionRequired`. The instructions that should be made clear, as described above, in the first quote?

Yes 👍 

cc @rkmanda

**Comment by bdefoy** [comment]
@konrad-jamrozik I made some edits to my previous comment based on feedback from Roopesh. In summary, the bot should:

- Add the `ARMChangesRequested` label when it sees a `NewApiVersionRequired` label
- Once the required changes have been made (`NewApiVersionRequired` label gets removed) or an exception has been granted (I believe there is another label added for the exception), add the `WaitForARMFeedback` label

**Comment by konrad-jamrozik** [comment]
@bdefoy can you check my understanding of how things work below? It goes a bit out of scope of this issue, but I am trying to wrap around my head around the broader workflow.

Here are some scenarios, based on your guidance and on [this diagram](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/909/API-spec-PR-review?anchor=management-plane-api-review):

### Scenario A, "ARM changes required, no new API version required", as currently implemented

Note the current implementation of scenario A is **also** how things **should** work.

1. A new PR is submitted.
2. Automation determines this is an ARM change, thus adding `ARMReview` label.
3. Automation determines there is no need for any new API version, so it **does not** add any `NewApiVersionRequired` label.
4. Automation adds the `WaitForARMFeedback` label.
5. ARM reviewer sees the `WaitForARMFeedback` label and thus starts a review. They leave comments and when done, **manually** add the label `ARMChangesRequested`.
6. Automation, upon seeing `ARMChangesRequested`, removes `WaitForARMFeedback`.
7. PR author sees the `ARMChangesRequested` label and thus starts addressing them. When done, they  **manually** remove the label `ARMChangesRequested`.
8. Automation, upon seeing removal of `ARMChangesRequested`, adds the label `WaitForARMFeedback`.
9. ARM reviewer sees the `WaitForARMFeedback` label and thus starts a review. Here we either go back to step 5, or, if there are no further remarks, the ARM reviewer **manually** adds the label `ARMSignedOff` and **manually** removes the label `WaitForARMFeedback`.
    - **Question**: should we automate removal of the `WaitForARMFeedback` label here? I see merged PRs that have both `ARMSignedOff` and `WaitForARMFeedback`, see [this query](https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Apr+label%3AArmSignedOff+label%3AWaitForArmFeedback+is%3Amerged).
    - In fact, I also see merged PRs that have `ARMSignedOff` and `ARMChangesRequested`, see [this query](https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Apr+label%3AArmSignedOff+label%3AArmChangesRequested+is%3Amerged), which I think is bug in our bots.

### Scenario B, "New ARM API version required", as it **should** work

1. A new PR is submitted.
2. Automation determines this is an ARM change, thus adding `ARMReview` label.
3. Automation determines there is a need for a new API version, so it adds a `NewApiVersionRequired` label.
4. Automation sees the `NewApiVersionRequired` label and thus adds `ARMChangesRequested` label. 
    - As a result, we end up in a situation like in Scenario A step 7, but now in addition we also have `NewApiVersionRequired` label.
    - **This is a new behavior, not currently implemented. This issue is about implementing it.**
5. The PR author sees the `NewApiVersionRequired` and `ARMChangesRequested` labels and starts addressing them.
6. As a result of PR author pushing commits with changes in new API instead of existing API, the automation removes the `NewApiVersionRequired` label.
    - Alternatively, the PR author may request a breaking-change-approval review. If approved by the breaking change reviewer, the reviewer **manually** adds the `Approved-BreakingChange` label.
7. PR author addresses the remaining changes requested by ARM, if any. When done, they **manually** remove the `ApiChangesRequested` label.
    - At this point the PR might either both `NewApiVersionRequired` and `Approved-BreakingChange` labels on it, or none of them.
8. Automation, upon seeing removal of `ARMChangesRequested`, adds the label `WaitForARMFeedback`.
9. Here we effectively end up in Scenario A, step 5, but with the possibility of the extra labels being present as elaborated in step 7.

### Scenario B, "New ARM API version required", as currently implemented

1. A new PR is submitted.
2. Automation determines this is an ARM change, thus adding `ARMReview` label.
3. Automation determines there is a need for a new API version, so it adds a `NewApiVersionRequired` label.
4. Automation adds the `WaitForARMFeedback` label.
    - **This is the problem described by this issue**. This label should not be added because `NewApiVersionRequired` label is present. Because of this, the steps 6-8 are necessary, but could be avoided.
5. The PR author sees the `NewApiVersionRequired` label and starts addressing it. 
6. **At the same time** an ARM reviewer sees the `WaitForARMFeedback` label and starts the review, just to realize they should hold off until `NewApiVersionRequired` is removed.
7. The ARM reviewer **manually** adds the `ARMChangesRequested` Label.
8. Automation, upon seeing `ARMChangesRequested`, removes `WaitForARMFeedback`.
9. As a result of PR author pushing commits with changes in new API instead of existing API (because of step 5), the automation removes the `NewApiVersionRequired` label.
    - Alternatively, the PR author may request a breaking-change-approval review. If approved by the breaking change reviewer, the reviewer **manually** adds the `Approved-BreakingChange` label.
10. Here we effectively end up in Scenario A, step 5, but with the possibility of the extra labels being present as elaborated in step 9.    

### Invariants (sanity checks)

There are some invariants that should always hold true. I am considering adding check in source code for them and emitting a warning if they are violated, as an additional method of verifying the logic is correct. Of course, only when it is possible to implement it with too much effort.

These invariants are to hold after the automation processed all labelling updates after given set of commits has been pushed to a PR.

- `NewApiVersionRequired` label should be added only by automation.
- `NewApiVersionRequired` label should be removed only by automation, triggered upon new commit push to the PR that fixes the issue.
- `WaitForARMFeedback` label should be added only by automation.
  - **Question**: Are there legitimate cases where a human should do it?
- `WaitForARMFeedback` label should be removed only by automation.
- `ARMChangesRequested` label should be added only by an ARM reviewer.
- `ARMChangesRequested` label should be removed only by the PR author.
- If `NewApiVersionRequired` label is present, the label `ARMChangesRequested` must be also present.
- `WaitForARMFeedback` label cannot be present at the same time as `NewApiVersionRequired` label.
- `WaitForARMFeedback` label cannot be present at the same time as `ARMChangesRequested` label.
- If `ARMSignedOff` label is present, `ARMChangesRequested` and `WaitForARMFeedback` labels cannot be present.
- If `ARMSignedOff` label is present, then labels `NewApiVersionRequired` and `Approved-BreakingChange` are either present both at the same time, or not at all.
- If the PR pertains to ARM, the `ARMReview` label must be always present.
- If the `ARMReview` label is present, at any point of time, exactly one of these labels must be present: `WaitForARMFeedback`, `ARMChangesRequested`, `ARMSignedOff`.

**Comment by konrad-jamrozik** [comment]
@bdefoy @weshaggard  
- https://github.com/Azure/azure-rest-api-specs/pull/24058

**Comment by konrad-jamrozik** [comment]
@bdefoy re `BreakingChangeReviewRequired`, I found [this code snippet](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/openapi-alps?path=public/rest-api-specs-scripts/src/prSummary.ts&version=GC02e956adcdd563ab5e0f43e60cf1cd259669cb85&line=744&lineEnd=747&lineStartColumn=1&lineEndColumn=6&_a=contents):

``` typescript
    // If exists both  "BreakingChangeReviewRequired" and "NewApiVersionRequired",remove the "NewApiVersionRequired"
    if (BreakingChangeLabels.every((l) => labelToEmit.includes(l))) {
      labelToEmit = labelToEmit.filter((l) => l !== "NewApiVersionRequired");
    }
```

The diagram says:

![image](https://github.com/Azure/azure-sdk-tools/assets/4429827/5a95bcdb-aa10-4c3a-98b0-eb82302ee0fa)

I am going to investigate what is the workflow behind that label and update it. But looks like it needs to be taken into account here.

**Comment by konrad-jamrozik** [comment]
> @konrad-jamrozik Konrad Jamrozik FTE I made some edits to my previous comment based on feedback from Roopesh. In summary, the bot should:
> 
> * Add the `ARMChangesRequested` label when it sees a `NewApiVersionRequired` label
> * Once the required changes have been made (`NewApiVersionRequired` label gets removed) or an exception has been granted (I believe there is another label added for the exception), add the `WaitForARMFeedback` label

@bdefoy I think we should add the following:

- These proposed rules apply only if the PR has `ARMReview`. I.e. they do not apply to `data-plane` PRs.
- `BreakingChangeReviewRequired` label should have the same treatment as `NewApiVersionRequired` label, per my 5/23/2023 update to [this comment](https://github.com/Azure/azure-sdk-tools/issues/6076#issuecomment-1553768166).
- If `NewApiVersionRequired` or `BreakingChangeReviewRequired` is present, in a PR with `ARMReview`, then we should ensure `WaitForARMFeedback` is not, unless `Approved-BreakingChange` is present.
  - By "ensure" I mean: don't add it if unnecessary, and remove it if somebody adds the label (we cannot prevent addition of a label, but we can remove it). I elaborated a bit on the challenge of doing the latter [here](https://github.com/Azure/azure-sdk-tools/issues/5311#issuecomment-1558171387).
  - When it comes to adding `WaitForARMFeedback` label it should happen through the existing rule which says the label gets added when the PR author manually removes the `ARMChangesRequested` label.

Do you agree with the above?

**Comment by konrad-jamrozik** [comment]
Our code handling labelling logic after the PR CI checks run (i.e. the `Summary` task) says that the `WaitForARMFeedback` label will be added when **all** of the following conditions hold:

From [`processARMReview`](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/openapi-alps?path=public/rest-api-specs-scripts/src/prSummary.ts&version=GC02e956adcdd563ab5e0f43e60cf1cd259669cb85&line=491&lineStartColumn=16&lineEndColumn=32&_a=contents):

- the PR introduces a new API
- the PR is labeled `resource-manager`
  - this is determined by [`getPRType`](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/openapi-alps?path=public/rest-api-specs-scripts/src/prSummary.ts&version=GC02e956adcdd563ab5e0f43e60cf1cd259669cb85&line=171&lineStartColumn=16&lineEndColumn=25&_a=contents) based on file paths edited by the PR
- the PR is not a draft

From [`isOkToAddWaitForARMFeedback`](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/openapi-alps?path=public/rest-api-specs-scripts/src/prSummary.ts&version=GC02e956adcdd563ab5e0f43e60cf1cd259669cb85&line=441&lineStartColumn=16&lineEndColumn=43&_a=contents):

- The PR has none of the following labels already: `WaitForARMFeedback`, `ARMSignedOff`, `ARMChangesRequested`
- The PR is not _pending breaking change review_. Where we say a PR is _pending breaking change review_ when it has label `BreakingChangeReviewRequired` and does not have label `Approved-BreakingChange`

**Comment by konrad-jamrozik** [comment]
https://devdiv.visualstudio.com/DevDiv/_git/openapi-alps/pullrequest/474083

@weshaggard @rkmanda @bdefoy please have a look at the PR, thx! :)
